signer/kms/aws/client.go

```diff
@@ -0,0 +1,341 @@
+		return &cmk, nil
+	}
+	return nil, lerr
+}
+
+// At the moment this function lies unused, but it is here for future if necessary
+func (a *awsClient) verify(ctx context.Context, sig, message io.Reader) error {
+	cmk, err := a.getCMK(ctx)
+	if err != nil {
+		return err
+	}
+
+	verifier, err := cmk.Verifier()
+	if err != nil {
+		return err
+	}
+
+	s, err := io.ReadAll(sig)
+	if err != nil {
+		return err
+	}
+
+	return verifier.Verify(message, s)
+}
+
+func (a *awsClient) verifyRemotely(ctx context.Context, sig, digest []byte) error {
+	cmk, err := a.getCMK(ctx)
+	if err != nil {
+		return err
+	}
+
+	alg := cmk.KeyMetadata.SigningAlgorithms[0]
+	messageType := types.MessageTypeDigest
+	if _, err := a.client.Verify(ctx, &akms.VerifyInput{
+		KeyId:            &a.keyID,
+		Message:          digest,
+		MessageType:      messageType,
+		Signature:        sig,
+		SigningAlgorithm: alg,
+	}); err != nil {
+		return fmt.Errorf("unable to verify signature: %w", err)
+	}
+
+	return nil
+}
+
+func (a *awsClient) sign(ctx context.Context, digest []byte, _ crypto.Hash) ([]byte, error) {
+	cmk, err := a.getCMK(ctx)
+	if err != nil {
+		return nil, err
+	}
+	alg := cmk.KeyMetadata.SigningAlgorithms[0]
+
+	messageType := types.MessageTypeDigest
+	out, err := a.client.Sign(ctx, &akms.SignInput{
+		KeyId:            &a.keyID,
+		Message:          digest,
+		MessageType:      messageType,
+		SigningAlgorithm: alg,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("signing with kms: %w", err)
+	}
+	return out.Signature, nil
+}
+
+func (a *awsClient) fetchPublicKey(ctx context.Context) (crypto.PublicKey, error) {
+	out, err := a.client.GetPublicKey(ctx, &akms.GetPublicKeyInput{
+		KeyId: &a.keyID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting public key: %w", err)
+	}
+	key, err := x509.ParsePKIXPublicKey(out.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("parsing public key: %w", err)
+	}
+	return key, nil
+}
+
+func (a *awsClient) fetchKeyMetadata(ctx context.Context) (*types.KeyMetadata, error) {
+	out, err := a.client.DescribeKey(ctx, &akms.DescribeKeyInput{
+		KeyId: &a.keyID,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting key metadata: %w", err)
+	}
+	return out.KeyMetadata, nil
+}
+
+type cmk struct {
+	KeyMetadata *types.KeyMetadata
+	PublicKey   crypto.PublicKey
+}
+
+func (c *cmk) HashFunc() crypto.Hash {
+	switch c.KeyMetadata.SigningAlgorithms[0] {
+	case types.SigningAlgorithmSpecRsassaPssSha256, types.SigningAlgorithmSpecRsassaPkcs1V15Sha256, types.SigningAlgorithmSpecEcdsaSha256:
+		return crypto.SHA256
+	case types.SigningAlgorithmSpecRsassaPssSha384, types.SigningAlgorithmSpecRsassaPkcs1V15Sha384, types.SigningAlgorithmSpecEcdsaSha384:
+		return crypto.SHA384
+	case types.SigningAlgorithmSpecRsassaPssSha512, types.SigningAlgorithmSpecRsassaPkcs1V15Sha512, types.SigningAlgorithmSpecEcdsaSha512:
+		return crypto.SHA512
+	default:
+		return 0
+	}
+}
+
+func (c *cmk) Verifier() (cryptoutil.Verifier, error) {
```

Comment:
I think most of this function can just be replaced with a call to [cryptoutil.NewVerifier](https://github.com/in-toto/go-witness/blob/1f9f95edcb8108cdcf8a1f94d723688e88919126/cryptoutil/verifier.go#L67).

This switch case seemingly boils down to a type assertion of c.PublicKey and instantiating the appropriate cryptoutil.Verifier depending on the results -- which is what NewVerifier does.

So I think you can just do `return cryptoutil.NewVerifier(c.PublicKey, cryptoutil.VerifyWithHash(c.HashFunc())`

Reply:
yeah the only reason that I kept this the way that it is is the fact that this is checking against the AWS types directly. Maybe this is simply unnecessary, provided that PKCS #1 v1.5 keys work as they should? I'll check it out

Reply:
yeah okay looks like we're all good here. Just ran and everything *seems* happy. Although not sure if we will hit some unexpected edge cases.

Reply:
Do you think it would be worth adding some steps with different signer / verifier types @mikhailswift ?